src/database/migrations/0031-create-sources.sql

```diff
@@ -0,0 +1,43 @@
+CREATE TYPE source_type AS ENUM ('data_provider', 'funder', 'organization', 'system');
+
+CREATE TABLE sources (
+  id INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
+  source_type source_type NOT NULL,
+  label VARCHAR NOT NULL,
+  organization_id INTEGER REFERENCES organizations(id),
+  funder_id INTEGER REFERENCES funders(id),
+  data_provider_id INTEGER REFERENCES data_providers(id),
+  created_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW(),
+  UNIQUE(organization_id, funder_id, data_provider_id),
```

Comment:
I don't think this does what you want it to do! `NULL`s are considered unequal to other `NULL`s, so two rows containing `NULL`s will always be unique.

```sql
=> create temporary table t (a integer, b integer, unique (a, b));
CREATE TABLE
=> insert into t (a, b) values (null, null);
INSERT 0 1
=> insert into t (a, b) values (null, null);
INSERT 0 1
=> insert into t (a, b) values (1, null);
INSERT 0 1
=> insert into t (a, b) values (1, null);
INSERT 0 1
=> insert into t (a, b) values (1, 2);
INSERT 0 1
=> insert into t (a, b) values (1, 2);
ERROR:  duplicate key value violates unique constraint "t_a_b_key"
DETAIL:  Key (a, b)=(1, 2) already exists.
```

I believe you need to have separate `UNIQUE` constraints on each column - assuming I'm following the intention correctly.

Relatedly, it would be good to have tests around duplicates!

(this is a follow-up to [an earlier conversation](https://github.com/PhilanthropyDataCommons/service/pull/1116/files#r1710630347))

Reply:
Today I learned this!  I could have sworn I've been told I couldn't enter a NULL column due to a unique constraint, but that must have been some kind of wild implanted memory.

Thanks, will fix up.

Reply:
Yikes. Nice catch, @jasonaowen!

@slifty We are presently on [14](https://github.com/PhilanthropyDataCommons/deploy/blob/main/compose.yml#L57) but this capability seems to require 15 or later. Perhaps we are overdue for upgrade anyway and this is one more reason to do it.

Reply:
Ah I was wondering about where our psql version was pinned!

I would love to be able to use it, but in the short term I could just do jason's fix (the only problem is I *think* it would allow more than one system / all-null source, so if we don't upgrade I'd want to add some constraint to somehow catch that case)

Reply:
Just had another thought -- I will talk with @kfogel .

There may be a use case for wanting more than one source for a given relationship.  For instance: The foundation foundation may want two sources, one to represent their GMS integration, another to represent their staff manually going in and editing data.

If that's the case this entire constraint probably goes away and the whole point is moot for now!